usr.sbin/inetd/parse_v2.c

```diff
@@ -0,0 +1,1067 @@
+#endif
+
+static invoke_result	parse_invoke_handler(bool *, char **, struct servtab *);
+static bool fill_default_values(struct servtab *);
+static bool parse_quotes(char **);
+static bool	skip_whitespace(char **);
+static int	size_to_bytes(char *arg);
+static bool infer_protocol_ip_version(struct servtab *);
+static bool	setup_internal(struct servtab *);
+static void	try_infer_socktype(struct servtab *);
+#ifdef IPSEC
+static void	setup_ipsec(struct servtab *);
+#endif
+
+/* v2 Key handlers infrastructure */
+
+/* v2 syntax Handler function, which must parse all values for its key */
+typedef hresult (*key_handler_func)(struct servtab *, vlist);
+
+/* List of v2 syntax key handlers */
+static struct key_handler {
+	const char *name;
+	key_handler_func handler;
+} key_handlers[] = {
+	{"bind", bind_handler},
+	{"socktype", socket_type_handler},
+	{"acceptfilter", filter_handler},
+	{"protocol", protocol_handler},
+	{"sndbuf", send_buf_handler},
+	{"recvbuf", recv_buf_handler},
+	{"wait", wait_handler},
+	{"service_max", service_max_handler},
+	{"user", user_handler},
+	{"group", group_handler},
+	{"exec", exec_handler},
+	{"args", args_handler},
+	{"ip_max", ip_max_handler},
+#ifdef IPSEC
+	{"ipsec", ipsec_handler}
+#endif
+};
+
+#define STRINGIFY(x) #x
+#define TOSTRING(x) STRINGIFY(x)
+
+/* Error Not Initialized */
+#define ENI(key) ERR("Required option '%s' not specified", (key))
+
+#define WAIT_WRN "Option 'wait' for internal service '%s' should be '%s'"
+
+/* Too Few Arguemnts (values) */
+#define TFA(key) ERR("Option '%s' has too few arguments", (key))
+
+/* Too Many Arguments (values) */
+#define TMA(key) ERR("Option '%s' has too many arguments", (key))
+
+/* Too Many Definitions */
+#define TMD(key) ERR("Option '%s' is already specified", (key))
+
+#define VALID_SOCKET_TYPES "stream, dgram, rdm, seqpacket, raw"
+
+parse_v2_result
+parse_syntax_v2(struct servtab *sep, char **cpp)
+{
+
+	/* Catch multiple semantic errors instead of skipping after one */
+	bool is_valid_definition = true;
+	/* Line number of service for error logging. */
+	size_t line_number_start = line_number;
+	invoke_result result;
+
+	if (debug) {
+		fprintf(stderr, "Found key-values definition line %zu: '%s'\n", 
+			line_number_start, sep->se_service);
+	}
+
+	for (;;) {
+		switch(result = 
+			parse_invoke_handler(&is_valid_definition, cpp, sep)) {
+		case INVOKE_SUCCESS:
+			/* Keep reading more options in. */
+			continue;
+		case INVOKE_FINISH:
+			/* Found a semicolon, do final checks and defaults and return. */
+			/* Skip whitespace after semicolon to end of line. */
+			while (**cpp == ' ' || **cpp == '\t') {
+				(*cpp)++;
+			}
+
+			if (is_valid_definition && fill_default_values(sep)) {
+				if (**cpp == '\0') {
+					*cpp = nextline(fconfig);
+				}
+				return V2_SUCCESS;
+			} else {
+				if (debug) {
+					fprintf(stderr, "Invalid definition, ignoring\n");
+				}
+				syslog(LOG_ERR, CONF_ERROR_FMT "Ignoring invalid definition.",
```

Comment:
Why doesn't this syslog use the macro?

Reply:
It uses line_number_start so the error will point to the first line of the service definition instead of the last line. Not sure if I should make another macro that leaves the line number as a variadic parameter for use in this scenario.